[draft] SEP client id metadata document (#1296)
modelcontextprotocol/modelcontextprotocol: docs/specification/draft/basic/authorization.mdx

### Client ID Metadata Documents

MCP clients and authorization servers **SHOULD** support OAuth Client ID Metadata Documents as specified in
[OAuth Client ID Metadata Document](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-client-id-metadata-document-00).
This approach enables clients to use HTTPS URLs as client identifiers, where the URL points to a JSON document
containing client metadata. This addresses the common MCP scenario where servers and clients have
no pre-existing relationship.

#### Implementation Requirements

MCP implementations supporting Client ID Metadata Documents **MUST** follow the requirements specified in
[OAuth Client ID Metadata Document](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-client-id-metadata-document-00).
Key requirements include:

**For MCP Clients:**

- Clients **MUST** host their metadata document at an HTTPS URL following RFC requirements
- The `client_id` URL **MUST** use the "https" scheme and contain a path component
- The metadata document **MUST** include at minimum: `client_id`, `client_name`, `redirect_uris`
- Clients **MUST** ensure the `client_id` value in the metadata matches the document URL exactly

> - Clients **MAY** use `private_key_jwt` for token endpoint authentication with appropriate JWKS configuration

Reviewer: ```suggestion
- Clients **MAY** use `private_key_jwt` for token endpoint authentication with appropriate JWKS configuration as described in [Section 6.2 of Client ID Metadata Document](https://www.ietf.org/archive/id/draft-ietf-oauth-client-id-metadata-document-00.html#section-6.2)
```

  "client_name": "Example MCP Client",
  "client_uri": "https://app.example.com",
  "logo_uri": "https://app.example.com/logo.png",
  "redirect_uris": [
    "http://127.0.0.1:3000/callback",
    "http://localhost:3000/callback"
  ],
  "grant_types": ["authorization_code"],
  "response_types": ["code"],
  "token_endpoint_auth_method": "none"
}
```

#### Client ID Metadata Documents Flow

The following diagram illustrates the complete flow when using Client ID Metadata Documents:

```mermaid
sequenceDiagram
    participant User
    participant Client as MCP Client
    participant Server as Authorization Server
    participant Metadata as Metadata Endpoint<br/>(Client's HTTPS URL)
    participant Resource as MCP Server

    Note over Client,Metadata: Client hosts metadata at<br/>https://app.example.com/oauth/metadata.json

    User->>Client: Initiates connection to MCP Server
    Client->>Server: Authorization Request<br/>client_id=https://app.example.com/oauth/metadata.json<br/>redirect_uri=http://localhost:3000/callback

    Note over Server: Authenticates user

    Note over Server: Detects URL-formatted client_id

> 

Reviewer: The user authentication step is missing from the diagram, do we want to include it here (before the metadata fetch) to demonstrate gating the fetch on the user being logged in?

    Server->>Metadata: GET https://app.example.com/oauth/metadata.json
    Metadata-->>Server: JSON Metadata Document<br/>{client_id, client_name, redirect_uris, ...}

>     Note over Server: Validates:<br/>1. client_id matches URL<br/>2. redirect_uri in allowed list<br/>3. Document structure valid<br/>4. Domain allowed via trust policy

Reviewer: ```suggestion
    Note over Server: Validates:<br/>1. client_id matches URL<br/>2. redirect_uri in allowed list<br/>3. Document structure valid<br/>4. (Optional) Domain allowed via trust policy
```

    Note over C,M: MCP communication continues with valid token
```

## Client Registration Approaches

MCP supports three client registration mechanisms. Choose based on your scenario:

> - **Client ID Metadata Documents**: When client and server have no prior relationship (most common)

Reviewer 2: how are clients choosing the approach to use? is there any concept of priority?

- **Client ID Metadata Documents**: When client and server have no prior relationship (most common)
- **Dynamic Client Registration**: For backwards compatibility or specific requirements

### Client ID Metadata Documents

MCP clients and authorization servers **SHOULD** support OAuth Client ID Metadata Documents as specified in
[OAuth Client ID Metadata Document](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-client-id-metadata-document-00).
This approach enables clients to use HTTPS URLs as client identifiers, where the URL points to a JSON document
containing client metadata. This addresses the common MCP scenario where servers and clients have
no pre-existing relationship.

#### Implementation Requirements

MCP implementations supporting Client ID Metadata Documents **MUST** follow the requirements specified in
[OAuth Client ID Metadata Document](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-client-id-metadata-document-00).
Key requirements include:

**For MCP Clients:**

- Clients **MUST** host their metadata document at an HTTPS URL following RFC requirements

> - The `client_id` URL **MUST** use the "https" scheme and contain a path component

Reviewer: ```suggestion
- The `client_id` URL **MUST** use the "https" scheme and contain a path component, e.g. `https://example.com/client.json`
```

    Note over C,M: MCP communication continues with valid token
```

### Client Registration Approaches

MCP supports three client registration mechanisms. Choose based on your scenario:

- **Pre-registration**: When client and server have an existing relationship
- **Client ID Metadata Documents**: When client and server have no prior relationship (most common)
- **Dynamic Client Registration**: For backwards compatibility or specific requirements

> ### Client ID Metadata Documents

Author: ```suggestion
#### Client ID Metadata Documents
```

    Note over Server: Detects URL-formatted client_id

    Server->>Metadata: GET https://app.example.com/oauth/metadata.json
    Metadata-->>Server: JSON Metadata Document<br/>{client_id, client_name, redirect_uris, ...}

    Note over Server: Validates:<br/>1. client_id matches URL<br/>2. redirect_uri in allowed list<br/>3. Document structure valid<br/>4. Domain allowed via trust policy

    alt Validation Success
        Server->>User: Display consent page with client_name
        User->>Server: Approves access
        Server->>Client: Authorization code via redirect_uri
        Client->>Server: Exchange code for token<br/>client_id=https://app.example.com/oauth/metadata.json
        Server-->>Client: Access token
        Client->>Resource: MCP requests with access token
        Resource-->>Client: MCP responses
    else Validation Failure
        Server->>User: Error response<br/>error=invalid_client or invalid_request
    end

    Note over Server: Cache metadata for future requests<br/>(respecting HTTP cache headers)
```

#### Discovery

Authorization servers advertise support via their metadata:

```json
{
  "client_id_metadata_document_supported": true
}
```

> MCP clients **SHOULD** check for this capability and may fall back to Dynamic Client Registration

Author: ```suggestion
MCP clients **SHOULD** check for this capability and **MAY** fall back to Dynamic Client Registration
```

Authorization servers **SHOULD** only automatically redirect the user agent if it trusts the redirection URI. If the URI is not trusted, the authorization server MAY inform the user and rely on the user to make the correct decision.

### Client ID Metadata Document Security

When implementing Client ID Metadata Documents, authorization servers **MUST** consider the security implications
detailed in [OAuth Client ID Metadata Document, Section 6](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-client-id-metadata-document-00#name-security-considerations).
Key considerations include:

#### Authorization Server Abuse Protection

The authorization server takes a URL as input from an unknown client and fetches that URL.

> A malicious client could use this to send non-metadata requests on behalf of the authorization server,

Reviewer: ```suggestion
A malicious client could use this to trigger the authorization server to make requests to arbitrary URLs,
```

Authorization servers **SHOULD** only automatically redirect the user agent if it trusts the redirection URI. If the URI is not trusted, the authorization server MAY inform the user and rely on the user to make the correct decision.

### Client ID Metadata Document Security

When implementing Client ID Metadata Documents, authorization servers **MUST** consider the security implications

> detailed in [draft-parecki-oauth-client-id-metadata-document-03 Section 7](https://www.ietf.org/archive/id/draft-parecki-oauth-client-id-metadata-document-03.txt).

Author: ```suggestion
detailed in [OAuth Client ID Metadata Document, Section 6](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-client-id-metadata-document-00#name-security-considerations).
```

Authorization servers **SHOULD** only automatically redirect the user agent if it trusts the redirection URI. If the URI is not trusted, the authorization server MAY inform the user and rely on the user to make the correct decision.

### Client ID Metadata Document Security

When implementing Client ID Metadata Documents, authorization servers **MUST** consider the security implications
detailed in [OAuth Client ID Metadata Document, Section 6](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-client-id-metadata-document-00#name-security-considerations).
Key considerations include:

#### Authorization Server Abuse Protection

The authorization server takes a URL as input from an unknown client and fetches that URL.
A malicious client could use this to trigger the authorization server to make requests to arbitrary URLs,
such as requests to private administration endpoints the authorization server has access to.

Authorization servers fetching metadata documents **MUST** protect against
[Server-Side Request Forgery (SSRF)](https://developer.mozilla.org/docs/Web/Security/Attacks/SSRF) attacks,
as well as against being used as Denial of Service (DoS) amplifiers:

> - Validate URLs and resolved IP addresses before fetching

Reviewer 2: what kind of validation are we talking about? block some ip ranges?

Key considerations include:

#### SSRF Protection

Authorization servers fetching metadata documents **MUST** protect against Server-Side Request Forgery (SSRF) attacks:

- Validate URLs and resolved IP addresses before fetching
- Limit response size (recommended 5 kilobytes)
- Implement request timeouts
- Rate limit outbound metadata fetches
- Never cache error responses or invalid documents

The authorization server takes a URL as input from an unknown client and fetches that URL.
A malicious client could use this to send non-metadata requests on behalf of the authorization server,
such as requests to private administration endpoints the authorization server has access to.

#### DDoS Protection

Authorization servers **MUST** protect against being used as DDoS amplifiers:

- Implement aggressive caching of metadata documents (respecting HTTP cache headers)
- Rate limit metadata fetch requests per client
- Monitor and alert on unusual metadata fetch patterns
- Only fetch client metadata after authenticating the user

> While there is no amplification in the fetch request bandwidth, aggressive caching minimizes

Reviewer 3: Since there is no DDoS amplification occurring, I think it would make more sense to call this whole section `DoS` instead of `DDoS`. Then you can move things like request timeouts and max response size from above into this section as well.

the risk of authorization servers being used in distributed denial of service attacks.

#### Localhost Redirect URI Risks

Client ID Metadata Documents cannot prevent `localhost` URL impersonation by themselves. An attacker can claim to be any client by:

1. Providing the legitimate client's metadata URL as their `client_id`
2. Binding to the any `localhost` port, and providing that address as the redirect_uri
3. Receiving the authorization code via the redirect when the user approves

The server will see the correct metadata document and the user will see the correct client name, making attack detection difficult.

Authorization servers:

- **SHOULD** display additional warnings for `localhost`-only redirect URIs
- **MAY** require additional attestation mechanisms for enhanced security

> - **MUST** clearly display the client's hostname during authorization

Reviewer 3: Is this the hostname of the Client ID Metadata URL or the hostname of the Redirect URI or the hostname of the Client URI?
